### PR TITLE
Fix trim double quotes for RestApiCallNode

### PR DESCRIPTION
**Pull request by AndriiLandiak**
## Pull Request description

_Issue:_ Currently, the implementation is not allowed to send a correct **plain/text** payload. The RestApiCallNode features a configuration field labeled "trimDoubleQuotes.", however, this approach has proven problematic. It does not adequately handle situations where specific symbols like newline characters \n, tabs \t, curly braces {}, square brackets [], and others are present in the serialized message.

_Fix:_ To address this issue, we should implement a more robust solution. Specifically, we should perform deserialization after the message has been serialized within the unbindMsg() method of the RuleNodeScriptEngine (JacksonUtil.toString()). 
This deserialization process should make use of JacksonUtil.fromString(). To ensure the correctness of this approach, comprehensive tests have been provided to cover scenarios involving these special symbols.

Rule node UI: https://github.com/thingsboard/thingsboard-rule-config-ui-ngx/pull/147

## General checklist

- [x] You have reviewed the guidelines [document](https://docs.google.com/document/d/1wqcOafLx5hth8SAg4dqV_LV3un3m5WYR8RdTJ4MbbUM/edit?usp=sharing).
- [ ] [Labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels#about-labels) that classify your pull request have been added.
- [ ] The [milestone](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/about-milestones) is specified and corresponds to fix version.  
- [x] Description references specific [issue](https://github.com/thingsboard/thingsboard/issues).
- [x] Description contains human-readable scope of changes.
- [x] Description contains brief notes about what needs to be added to the documentation.
- [x] No merge conflicts, commented blocks of code, code formatting issues.
- [x] Changes are backward compatible or upgrade script is provided.
- [x] Similar PR is opened for PE version to simplify merge. Crosslinks between PRs added. Required for internal contributors only.
  

## Back-End feature checklist

- [x] Added corresponding unit and/or integration test(s). Provide written explanation in the PR description if you have failed to add tests.
- [x] If new dependency was added: the dependency tree is checked for conflicts.
- [x] If new service was added: the service is marked with corresponding @TbCoreComponent, @TbRuleEngineComponent, @TbTransportComponent, etc.
- [x] If new REST API was added: the RestClient.java was updated, issue for [Python REST client](https://github.com/thingsboard/thingsboard-python-rest-client) is created.
- [x] If new yml property was added: make sure a description is added (above or near the property).

### CODE REVIEW

File: rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/rest/TbHttpClient.java [hunk 1]
```diff
@@ -242,12 +242,18 @@ public URI buildEncodedUri(String endpointUrl) {
...
     }
 
-    private String getData(TbMsg msg) {
-        String data = msg.getData();
+    private String getData(TbMsg tbMsg) {
+        return trimDoubleQuotesIfRequired(tbMsg.getData());
+    }
 
-        if (config.isTrimDoubleQuotes()) {
+    protected String trimDoubleQuotesIfRequired(String data) {
```
Comment (smatvienko-tb) [inline rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/rest/TbHttpClient.java:249]: 1. The feature has to be renamed to make more sense. as it acts as JSON.parse to make plain/text from JSON string.
2. UI and description and documentation have to be adjusted.
3. the method getData should get `getConfig().isTrimDoubleQuotes()`as a boolean parameter. This make your code better and much easier to test without excessive mocking

File: rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/rest/TbHttpClient.java [hunk 2]
```diff
@@ -242,12 +242,18 @@ public URI buildEncodedUri(String endpointUrl) {
         return uri;
     }
 
-    private String getData(TbMsg msg) {
-        String data = msg.getData();
+    private String getData(TbMsg tbMsg) {
```
Comment (smatvienko-tb) [inline rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/rest/TbHttpClient.java:245]: the method seems useless

File: rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/rest/TbHttpClientTest.java [hunk 3]
```diff
@@ -60,10 +61,12 @@ public class TbHttpClientTest {
 
     EventLoopGroup eventLoop;
     TbHttpClient client;
+    TbRestApiCallNodeConfiguration config;
 
     @Before
     public void setUp() throws Exception {
         client = mock(TbHttpClient.class);
+        config = mock(TbRestApiCallNodeConfiguration.class);
```
Comment (smatvienko-tb) [inline rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/rest/TbHttpClientTest.java:69]: I guess this is not necessary, just make a refactoring of the subject method

File: rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/rest/TbHttpClientTest.java [hunk 4]
```diff
@@ -224,4 +227,55 @@ public void testHeadersToMetaData() {
         Assertions.assertThat(data.get("Set-Cookie")).isEqualTo("[\"sap-context=sap-client=075; path=/\",\"sap-token=sap-client=075; path=/\"]");
     }
 
-}
\ No newline at end of file
+    @Test
```
Comment (smatvienko-tb) [inline rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/rest/TbHttpClientTest.java:230]: Please, make a parametrized test

**Review by smatvienko-tb — CHANGES_REQUESTED** [review]
Hey!
This fix makes perfect sense, as we discussed in the morning. 

But for the product, it requires more effort from You.
Please find my comments below.

**Comment by smatvienko-tb** [comment]
as a reference:

JSON.stringify()
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify

JSON.parse()
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse

**Comment by smatvienko-tb** [comment]
please, add in the description that the current implementation are not allowed to send a correct **plain/text** payload

**Comment by ashvayka** [comment]
Let's analyze the headers configured by the user. If we are sending the JSON ("Content-type" = "application/json;charset=UTF-8"), we should create an HTTPEntity with correct type and use it. We can also apply validation that we have a valid JSON before posting it. IF we are sending the request as plain/text then we should convert it to String. 
See more here: https://www.baeldung.com/spring-resttemplate-post-json#2-handling-character-encoding

File: rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/rest/TbHttpClientTest.java [hunk 5]
```diff
@@ -219,9 +225,41 @@ public void testHeadersToMetaData() {
...
-        Assertions.assertThat(data.get("Set-Cookie")).isEqualTo("[\"sap-context=sap-client=075; path=/\",\"sap-token=sap-client=075; path=/\"]");
+        Assertions.assertEquals(2, data.size());
+        Assertions.assertEquals(data.get("Content-Type"), "binary");
+        Assertions.assertEquals(data.get("Set-Cookie"), "[\"sap-context=sap-client=075; path=/\",\"sap-token=sap-client=075; path=/\"]");
     }
 
-}
\ No newline at end of file
+    @ParameterizedTest
+    @MethodSource("provideParameters")
```
Comment (smatvienko-tb) [inline rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/rest/TbHttpClientTest.java:234]: Please, use
`@ValueSource(strings = {"", "  "})`
for primitive types

File: rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/rest/TbHttpClient.java [hunk 6]
```diff
@@ -242,12 +242,16 @@ public URI buildEncodedUri(String endpointUrl) {
...
+        return parseJsonStringToPlainText(tbMsg.getData(), config.isParseToPlainText());
+    }
 
-        if (config.isTrimDoubleQuotes()) {
+    protected String parseJsonStringToPlainText(String data, boolean parseToJson) {
```
Comment (smatvienko-tb) [inline rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/rest/TbHttpClient.java:249]: it looks like the method doing something only for parseToJson == true
It makes sense to move such a case upper and not fire the method if parseToJson == false

Comment (smatvienko-tb) [inline rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/rest/TbHttpClient.java:249]: I see there is "no body" configuration present. This means parsing data makes no sense at all when "no body" == true

**Review by smatvienko-tb — COMMENTED** [review]
Please, find some minor notes below

**Pull request merged**